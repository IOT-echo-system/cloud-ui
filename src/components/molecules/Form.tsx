import {Stack, Typography} from '@mui/material'
import type {FormEvent} from 'react'
import React from 'react'
import type {FormInputType} from '../atoms'
import {Button, FormContainer, FormInput} from '../atoms'

export type FormProps = {
  handleSubmit: (event: FormEvent<HTMLFormElement>) => void
  title: string
  inputFields: FormInputType[]
  submitBtnText: string
  submitBtnDisabled?: boolean
  loading?: boolean
}

export const Form: React.FC<FormProps> = props => {
  return (
    <FormContainer onSubmit={props.handleSubmit}>
      <Stack spacing={2}>
        <Typography variant={'h5'}>{props.title}</Typography>
        {props.inputFields.map((inputField, index) => (
          <FormInput key={`input-${index}`} {...inputField} />
        ))}
        <Button
          loading={props.loading ?? false}
          type={'submit'}
          variant={'contained'}
          size={'large'}
          disabled={props.submitBtnDisabled}
        >
          {props.submitBtnText}
        </Button>
      </Stack>
    </FormContainer>
  )
}
