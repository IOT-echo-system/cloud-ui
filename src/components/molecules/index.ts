export * from './Header/Header'
export * from './Footer'
export * from './Form'
export * from './ValidatedProfile'
export * from './FetchDetails'
export * from './CustomModal'
export * from './BoardSecretKey/BoardSecretKey'
export * from './ConfirmationModals'
export * from './MqttSubscriber'
