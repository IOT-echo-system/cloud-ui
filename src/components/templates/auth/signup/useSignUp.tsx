import type {ChangeEvent} from 'react'
import type React from 'react'
import {useState} from 'react'
import {useRouter} from 'next/router'
import type {FormInputType} from '../../../atoms'
import {useResetPassword} from '../resetPassword/useResetPassword'
import {useForm, useToast} from '../../../../hooks'
import {UserService} from '../../../../services'
import {Config} from '../../../../config'

type UseSignUpReturnType = {
  inputFields: FormInputType[]
  handleSubmit: (event: React.FormEvent<HTMLFormElement>) => void
  submitBtnDisabled: boolean
  loading: boolean
}

export const useSignUp = (): UseSignUpReturnType => {
  const router = useRouter()
  const {values, onChange, handleSubmit} = useForm({name: '', email: ''})
  const {inputFields: passwordInputFields} = useResetPassword(false, '')
  const [loading, setLoading] = useState(false)
  const toast = useToast()

  const handleChange = <K extends keyof typeof values>(keyName: K) => {
    return (event: ChangeEvent<HTMLInputElement>) => {
      onChange(keyName, event.target.value)
    }
  }

  const onSubmit = () => {
    const password = passwordInputFields[0].value as string
    setLoading(true)
    UserService.registration({password, ...values})
      .then(() => {
        toast.success('Your account has been created successfully')
        return router.push(Config.LOGIN_PAGE_PATH)
      })
      .catch(toast.error)
      .finally(() => {
        setLoading(false)
      })
  }

  const inputFields: FormInputType[] = [
    {inputType: 'textField', value: values.name, onChange: handleChange('name'), label: 'Name', required: true},
    {
      inputType: 'textField',
      type: 'email',
      value: values.email,
      onChange: handleChange('email'),
      label: 'Email',
      required: true
    },
    ...passwordInputFields
  ]

  return {
    handleSubmit: handleSubmit(onSubmit),
    submitBtnDisabled: inputFields.some(inputField => inputField.error),
    inputFields,
    loading
  }
}
