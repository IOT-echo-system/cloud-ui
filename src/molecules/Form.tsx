import {Stack, Typography} from '@mui/material'
import type {FormEvent} from 'react'
import React from 'react'
import type {FormInputType} from '@/atoms'
import {Button, FormContainer, FormInput} from '@/atoms'

export type FormProps = {
  handleSubmit: (event: FormEvent<HTMLFormElement>) => void
  title: string
  error?: string
  inputFields: FormInputType[]
  submitBtnText: string
  submitBtnDisabled?: boolean
}

export const Form: React.FC<FormProps> = props => {
  return (
    <FormContainer onSubmit={props.handleSubmit}>
      <Stack spacing={2}>
        <Typography variant={'h5'}>{props.title}</Typography>
        {props.error && (
          <Typography variant={'body1'} color={'error'}>
            {props.error}
          </Typography>
        )}
        {props.inputFields.map((inputField, index) => (
          <FormInput key={`input-${index}`} {...inputField} />
        ))}
        <Button type={'submit'} variant={'contained'} size={'large'} disabled={props.submitBtnDisabled}>
          {props.submitBtnText}
        </Button>
      </Stack>
    </FormContainer>
  )
}
