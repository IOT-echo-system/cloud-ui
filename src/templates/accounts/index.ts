export * from './signup/SignUp'
export * from './login/Login'
