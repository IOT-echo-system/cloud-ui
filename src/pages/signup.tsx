import type {NextPage} from 'next'
import type {ChangeEvent} from 'react'
import React, {useState} from 'react'
import {Stack, TextField, Typography} from '@mui/material'
import {useForm, useSelector} from '@/hooks'
import {Button} from '@/components'
import WebClient from 'web-client-starter/lib'
import {useRouter} from 'next/router'
import {Config} from '@/config'
import type {ServerError} from '@/typing/error'
import { AuthFormContainer } from '@/modules/login/AuthFormContainer'

const SignUp: NextPage = () => {
  const site = useSelector(state => state.site)
  const router = useRouter()
  const [error, setError] = useState('')
  const [confPassword, setConfPassword] = useState('')
  const {values, onChange, handleSubmit} = useForm({name: '', email: '', password: ''})

  const handleChange = <K extends keyof typeof values>(keyName: K) => {
    return (event: ChangeEvent<HTMLInputElement>) => {
      onChange(keyName, event.target.value)
    }
  }

  const onSubmit = () => {
    setError('')
    WebClient.post({
      baseUrl: Config.BACKEND_BASE_URL,
      path: Config.SIGN_UP_PATH,
      body: values
    })
      .then(() => router.push('/login'))
      .catch((error: ServerError) => {
        setError(error.errorMessage)
      })
  }

  const handleConfPassword = (event: ChangeEvent<HTMLInputElement>): void => {
    setConfPassword(event.target.value)
  }

  const errorOnPassword = confPassword.length !== 0 && values.password !== confPassword

  return (
    <AuthFormContainer onSubmit={handleSubmit(onSubmit)} className={site.theme}>
      <Stack spacing={2}>
        <Typography variant={'h5'}>{site.title} Sign up</Typography>
        {error && (
          <Typography variant={'body1'} color={'error'}>
            {error}
          </Typography>
        )}
        <TextField
          value={values.name}
          onChange={handleChange('name')}
          label='Name'
          variant='outlined'
          required
        />
        <TextField
          type={'email'}
          value={values.email}
          onChange={handleChange('email')}
          label='Email'
          variant='outlined'
          required
        />
        <TextField
          type={'password'}
          value={values.password}
          onChange={handleChange('password')}
          label='Password'
          variant='outlined'
          required
        />
        <TextField
          type={'password'}
          value={confPassword}
          onChange={handleConfPassword}
          label='Confirm Password'
          variant='outlined'
          required
          error={errorOnPassword}
          helperText={errorOnPassword ? 'password and confirm password should match.' : ''}
        />
        <Button
          type={'submit'}
          variant={'contained'}
          size={'large'}
          disabled={values.password.length < 6 || values.password !== confPassword}
        >
          Sign up
        </Button>
      </Stack>
    </AuthFormContainer>
  )
}

export default SignUp
