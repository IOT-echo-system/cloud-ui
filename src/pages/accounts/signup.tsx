import type {NextPage} from 'next'
import React from 'react'
import {SignUp} from '@/templates/accounts'

const SignUpPage: NextPage = () => {
  return <SignUp />
}

export default SignUpPage
