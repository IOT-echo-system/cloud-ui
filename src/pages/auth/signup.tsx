import type {NextPage} from 'next'
import React from 'react'
import {SignUp} from '../../components/templates/auth'

const SignUpPage: NextPage = () => {
  return <SignUp />
}

export default SignUpPage
