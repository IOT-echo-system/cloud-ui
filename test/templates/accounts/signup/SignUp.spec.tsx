import React from 'react'
import {render} from '@testing-library/react'
import * as UseSignUpHooks from '../../../../src/templates/accounts/signup/useSignUp'
import {SignUp} from '../../../../src/templates/accounts'
import {Form} from '../../../../src/molecules/Form'

jest.mock('../../../../src/molecules/Form')
jest.mock('../../../../src/templates/accounts/signup/useSignUp')

describe('SignUp Template Test', () => {
  beforeEach(() => {
    jest.resetAllMocks()

    jest
      .spyOn(UseSignUpHooks, 'useSignUp')
      .mockReturnValue({inputFields: [], error: '', handleSubmit: jest.fn(), submitBtnDisabled: false})
  })

  afterEach(jest.resetAllMocks)

  it('should match with the snapshot', () => {
    const {container} = render(<SignUp />)

    expect(container).toMatchSnapshot()

    expect(UseSignUpHooks.useSignUp).toHaveBeenCalledTimes(1)
    expect(Form).toHaveBeenCalledTimes(1)
    expect(Form).toHaveBeenCalledWith(
      {
        error: '',
        handleSubmit: expect.any(Function),
        inputFields: [],
        submitBtnText: 'Sign up',
        title: 'Sign up',
        submitBtnDisabled: false
      },
      {}
    )
  })
})
