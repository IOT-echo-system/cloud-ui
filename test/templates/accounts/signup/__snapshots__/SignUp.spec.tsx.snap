// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp Template Test should match with the snapshot 1`] = `
<div>
  <div
    class="MuiStack-root css-1edlbwa-MuiStack-root"
  />
</div>
`;
